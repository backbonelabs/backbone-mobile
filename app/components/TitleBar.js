import React, { PropTypes } from 'react';
import { View, TouchableOpacity } from 'react-native';
import Icon from 'react-native-vector-icons/FontAwesome';
import HeadingText from './HeadingText';
import BodyText from './BodyText';
import styles from '../styles/titleBar';

const TitleBar = (props) => {
  const routeStack = props.navigator.getCurrentRoutes();
  const previousRoute = routeStack.length >= 2 ? routeStack[routeStack.length - 2] : null;
  const leftButton = previousRoute && props.currentRoute.showBackButton ? (
    <TouchableOpacity style={styles.leftComponent} onPress={props.navigator.pop}>
      <Icon
        name="angle-left"
        style={styles.buttonIcon}
        size={styles.$leftButtonIconSize}
        color={styles._buttonIcon.color}
      />
      <BodyText style={styles._sideContainersText} numberOfLines={1}>
        {previousRoute.title}
      </BodyText>
    </TouchableOpacity>
  ) : undefined;

  // The route's rightComponent, if present, will be passed the navigator object
  // so it can perform navigation functions
  const RightComponent = props.currentRoute.rightComponent;
  const rightButton = RightComponent ? (
    <View style={styles.rightComponent}>
      <RightComponent navigator={props.navigator} />
    </View>
  ) : undefined;

  // TitleBar will be visible, i.e., extend pass the status bar, only if the route has a title
  const titleBarStyles = props.currentRoute.title ? styles.visibleTitleBar : styles.hiddenTitleBar;

  return (
    <View style={[titleBarStyles, props.style]}>
      <View style={styles.sideContainers}>{leftButton}</View>
      <HeadingText size={2} style={styles._title}>{props.currentRoute.title}</HeadingText>
      <View style={styles.sideContainers}>{rightButton}</View>
    </View>
  );
};

TitleBar.propTypes = {
  navigator: PropTypes.object,
  currentRoute: PropTypes.shape({
    title: PropTypes.string,
    showBackButton: PropTypes.bool,
    rightComponent: PropTypes.func,
  }),
  style: PropTypes.object,
};

TitleBar.defaultProps = {
  currentRoute: {},
};

export default TitleBar;
