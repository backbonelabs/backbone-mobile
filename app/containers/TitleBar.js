import React, { PropTypes } from 'react';
import {
  Image,
  View,
  TouchableOpacity,
} from 'react-native';
import { connect } from 'react-redux';
import Icon from 'react-native-vector-icons/MaterialIcons';
import color from 'color';
import HeadingText from '../components/HeadingText';
import settingsIcon from '../images/settings-icon.png';
import routes from '../routes';
import { getColorHexForLevel } from '../utils/levelColors';
import styles from '../styles/titleBar';

const TitleBar = (props) => {
  const level = props.training.selectedLevelIdx;
  const levelColorCode = getColorHexForLevel(level);

  const routeStack = props.navigator.getCurrentRoutes();
  const previousRoute = routeStack.length >= 2 ? routeStack[routeStack.length - 2] : null;
  const leftButton = previousRoute && props.titleBar.showBackButton ? (
    <TouchableOpacity
      style={styles.leftComponent}
      onPress={props.disableBackButton ? null : props.navigator.pop}
    >
      <Icon
        name="keyboard-arrow-left"
        style={[
          styles.buttonIcon,
          {
            color: props.disableBackButton ?
              color(levelColorCode).clearer(0.6).rgbString() : levelColorCode,
          },
        ]}
        size={styles.$backButtonIconSize}
        color={getColorHexForLevel(level)}
      />
    </TouchableOpacity>
  ) : (
    <TouchableOpacity
      style={styles.leftComponent}
      onPress={() => props.navigator.push(routes.profile)}
    >
      {/* TODO: REPLACE WITH IMAGE COMPONENT OF USER'S PHOTO */}
      <Icon
        name="person"
        style={styles.profileIcon}
        color="#FFFFFF"
        size={styles.$profileIconSize}
      />
    </TouchableOpacity>
  );

  // The right component will be the settings icon by default, but can be
  // overridden by defining a rightComponent in the route config. The
  // route's rightComponent, if present, will be passed the navigator
  // object so it can perform navigation functions.
  const rightComponent = props.titleBar.rightComponent ? (
    <props.titleBar.rightComponent navigator={props.navigator} />
  ) : (
    <TouchableOpacity
      onPress={() => {
        props.navigator.push(routes.settings);
      }}
    >
      <Image
        name="settingsIcon" // primarily used for aiding testing
        source={settingsIcon}
        style={styles.icon}
      />
    </TouchableOpacity>
  );

  const rightButton = props.titleBar.showRightComponent && (
    <View style={styles.rightComponent}>
      {rightComponent}
    </View>
  );

  // TitleBar will be visible, i.e., extend pass the status bar, only if the route has a title
  const titleBarStyles = props.titleBar.title ? styles.visibleTitleBar : styles.hiddenTitleBar;
  const titleTextStyles = [
    styles._centerContainer,
    { color: getColorHexForLevel(level) },
  ];

  return (
    <View style={[titleBarStyles, props.titleBar.styles, props.style]}>
      <View style={styles.sideContainers}>{leftButton}</View>
      <HeadingText size={2} style={titleTextStyles} >{props.titleBar.title}</HeadingText>
      <View style={styles.sideContainers}>{rightButton}</View>
    </View>
  );
};

TitleBar.propTypes = {
  navigator: PropTypes.object,
  disableBackButton: PropTypes.bool,
  style: PropTypes.object,
  titleBar: PropTypes.shape({
    name: PropTypes.string,
    title: PropTypes.string,
    showBackButton: PropTypes.bool,
    showRightComponent: PropTypes.bool,
    rightComponent: PropTypes.func,
    styles: PropTypes.object,
  }).isRequired,
  training: PropTypes.shape({
    selectedLevelIdx: PropTypes.number,
  }).isRequired,
};

const mapStateToProps = (state) => {
  const { app: { titleBar }, training } = state;
  return { titleBar, training };
};

export default connect(mapStateToProps)(TitleBar);
