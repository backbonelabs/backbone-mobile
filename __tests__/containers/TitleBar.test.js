import React from 'react';
import { View } from 'react-native';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { asyncActionMiddleware } from 'redux-async-action';
import color from 'color';
import TitleBar from '../../app/containers/TitleBar';

describe('TitleBar Component', () => {
  const defaultState = {
    app: {
      titleBar: {
        name: null,
        title: null,
        component: null,
        showLeftComponent: false,
        showRightComponent: false,
        showNavbar: false,
        rightComponent: null,
      },
    },
    training: {
      selectedLevelIdx: 0,
    },
    user: {
      user: {
        nickname: 'testNickname',
      },
    },
  };
  const currentRoute = {
    name: 'testing',
  };

  const navigatorWithOneRoute = {
    getCurrentRoutes() {
      return [{}];
    },
  };
  const navigatorWithTwoRoutes = {
    getCurrentRoutes() {
      return [{}, {}];
    },
  };
  const configuredStore = configureStore([asyncActionMiddleware, thunk]);

  test('renders as expected', () => {
    const store = configuredStore(defaultState);
    const wrapper = shallow(
      <TitleBar navigator={navigatorWithOneRoute} currentRoute={currentRoute} />,
      { context: { store } },
    ).dive();

    expect(wrapper).toMatchSnapshot();
  });

  test('shows a custom left component that receives navigator/nickname as a prop', () => {
    const leftComponent = () => <View />;
    const store = configuredStore({
      ...defaultState,
      app: {
        ...defaultState.app,
        titleBar: {
          ...defaultState.app.titleBar,
          showLeftComponent: true,
          leftComponent,
        },
      },
    });
    const wrapper = shallow(
      <TitleBar navigator={navigatorWithOneRoute} currentRoute={currentRoute} />,
      { context: { store } },
    ).dive();

    const LeftComponentWrapper = wrapper.childAt(0).childAt(0).childAt(0);
    expect(LeftComponentWrapper.name()).toBe('leftComponent');
    expect(LeftComponentWrapper.props()).toHaveProperty('navigator', navigatorWithOneRoute);
    expect(LeftComponentWrapper.props())
      .toHaveProperty('nickname', defaultState.user.user.nickname);
  });

  test('disables the back button', () => {
    const store = configuredStore({
      ...defaultState,
      app: {
        ...defaultState.app,
        titleBar: {
          ...defaultState.app.titleBar,
          showLeftComponent: true,
        },
      },
    });
    const wrapper = shallow(
      <TitleBar navigator={navigatorWithTwoRoutes} disableBackButton currentRoute={currentRoute} />,
      { context: { store } },
    ).dive();
    const touchableWrapper = wrapper.childAt(0).childAt(0);
    const iconWrapper = wrapper.find('Icon').at(0);
    expect(touchableWrapper).toBeDefined();
    expect(iconWrapper).toBeDefined();
    expect(iconWrapper.props()).toHaveProperty('name', 'keyboard-arrow-left');
    expect(iconWrapper.props().style).toEqual(expect.arrayContaining([{
      color: color('rgba(0, 0, 0, 0.54)').clearer(0.6).rgbString(),
    }]));
  });

  test('uses the correct color style', () => {
    // First test with level index 0
    let store = configuredStore({
      ...defaultState,
      app: {
        ...defaultState.app,
        titleBar: {
          ...defaultState.app.titleBar,
          showLeftComponent: true,
        },
      },
    });
    let wrapper = shallow(
      <TitleBar navigator={navigatorWithTwoRoutes} currentRoute={currentRoute} />,
      { context: { store } },
    ).dive();
    let expectedStyles = [{ color: 'rgba(0, 0, 0, 0.54)' }];
    let leftProps = wrapper.find('Icon').at(0).props();
    let centerProps = wrapper.childAt(1).props();
    expect(leftProps.style).toEqual(expect.arrayContaining(expectedStyles));
    expect(centerProps.style).toEqual(expect.arrayContaining(expectedStyles));

    // Test with level index 2
    const newLevelIdx = 2;
    store = configuredStore({
      ...defaultState,
      app: {
        ...defaultState.app,
        titleBar: {
          ...defaultState.app.titleBar,
          showLeftComponent: true,
        },
      },
      training: {
        ...defaultState.training,
        selectedLevelIdx: newLevelIdx,
      },
    });
    wrapper = shallow(
      <TitleBar navigator={navigatorWithTwoRoutes} currentRoute={currentRoute} />,
      { context: { store } },
    ).dive();
    expectedStyles = [{ color: 'rgba(0, 0, 0, 0.54)' }];
    leftProps = wrapper.find('Icon').at(0).props();
    centerProps = wrapper.childAt(1).props();
    expect(leftProps.style).toEqual(expect.arrayContaining(expectedStyles));
    expect(centerProps.style).toEqual(expect.arrayContaining(expectedStyles));
  });

  test('can properly hide the right component', () => {
    const store = configuredStore(defaultState);
    const wrapper = shallow(
      <TitleBar navigator={navigatorWithOneRoute} currentRoute={currentRoute} />,
      { context: { store } },
    ).dive();
    expect(wrapper.childAt(2).children()).toHaveLength(0);
  });

  test('shows the settings icon', () => {
    const store = configuredStore({
      ...defaultState,
      app: {
        ...defaultState.app,
        titleBar: {
          ...defaultState.app.titleBar,
          showRightComponent: true,
        },
      },
    });
    const wrapper = shallow(
      <TitleBar navigator={navigatorWithOneRoute} currentRoute={currentRoute} />,
      { context: { store } },
    ).dive();
    const touchableWrapper = wrapper.childAt(2).childAt(0).childAt(0);

    expect(touchableWrapper.name()).toBe('TouchableOpacity');
    expect(touchableWrapper.childAt(0).name()).toBe('Image');
    expect(touchableWrapper.childAt(0).props()).toHaveProperty('name', 'settingsIcon');
  });

  test('shows a custom right component that receives navigator as a prop', () => {
    const RightComponent = () => <View />;
    const store = configuredStore({
      ...defaultState,
      app: {
        ...defaultState.app,
        titleBar: {
          ...defaultState.app.titleBar,
          showRightComponent: true,
          rightComponent: RightComponent,
        },
      },
    });
    const wrapper = shallow(
      <TitleBar navigator={navigatorWithOneRoute} currentRoute={currentRoute} />,
      { context: { store } },
    ).dive();

    const rightComponentWrapper = wrapper.childAt(2).childAt(0).childAt(0);
    expect(rightComponentWrapper.name()).toBe('RightComponent');
    expect(rightComponentWrapper.props()).toHaveProperty('navigator', navigatorWithOneRoute);
  });
});

