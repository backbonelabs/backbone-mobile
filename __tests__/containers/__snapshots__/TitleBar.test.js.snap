// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TitleBar Component renders as expected 1`] = `
<View
  style={
    Array [
      undefined,
      undefined,
      undefined,
    ]
  }
>
  <View />
  <HeadingText
    allowFontScaling={false}
    size={3}
    style={
      Array [
        undefined,
        Object {
          "color": "rgba(0, 0, 0, 0.54)",
        },
      ]
    }
  />
  <View />
</View>
`;
